util/argo/audit_logger.go

```diff
@@ -0,0 +1,104 @@
+package argo
+
+import (
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+
+	"fmt"
+	"time"
+
+	"github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
+)
+
+type AuditLogger struct {
```

Comment:
What tests do you think need writing for this? Unit tests? An e2e test?

Reply:
I was thing if you change this to an interface, I think you'll be able to create a mock using `mockery` and easily test it. But I think you can actually better test this using the Kubernetes fakes elsewhere.

Reply:
I didn't think about testing, it's my second Go PR ever so any guidance is accepted :) I guess it would be great to watch events in the different cases and see how they get published.

Reply:
For a unit test, I think there is nothing special to do, just standard `go test` tests. The way that I think about test cases is this, is there some ternary logic? I.e. does something happen under some circumstance but no others? I.e. is there `if` statements? These are best tested using unit tests.

So I'm thinking we probably want these test cases:

* We send events when ...
  * The spec is invalid (operator.go#L166)
  * Timeout (L187)
  * Unable to load artifactory config  (L213)
  * Params substitution error (L223)
  * PVC creation error (L232)
  *  Entry template execution (L243)
  * PVC deletion error (L287)
  * Node warning/failure/error/success (L300/L303/L307/L310)

Reply:
Hi @alexec, I was able to create unit tests for 
- The spec is invalid (operator.go#L166)
- Timeout (L187)
- Unable to load artifactory config (L213)

I wasn't able to create the test for `Params substitution error (L223)` as the code path is very tricky to reach (an invalid parameter is automatically interpreted as string without returning error) Ref: [operator.go#L205](https://github.com/argoproj/argo/blob/master/workflow/controller/operator.go#L205) -> [operator.go#L2129](https://github.com/argoproj/argo/blob/master/workflow/controller/operator.go#L2129) -> [util.go#L378-L379](https://github.com/argoproj/argo/blob/071eb1126cae351c5059246d2c375dc47175a3d6/workflow/common/util.go#L378-L379)

The other 4 tests I suspect require e2e because they actually need stuff running and might also be quite hard to reproduce. I'll think more about it

Reply:
I started adding some e2e tests from https://github.com/argoproj/argo/pull/1930/commits/8621536484a8eed1b3c38ba4b8c2c92fce7624b1 but let me know if i'm going the wrong direction

Reply:
Hi @alexec , do you think the test I added are sufficient? The remaining uncovered cases (which i'm not sure i can reproduce) are:
- Params substitution error (read comment above)
- Exit template execution error (seems to either fail validation or fail the the workflow, i wasn't able to trigger an error instead)
- PVC deletion error
- Node error (i did failure and success)

Can you point me to some reproducing template for those? Otherwise is the coverage enough? 

Thanks in advance for your time and help! :)
Alessandro

Reply:
I think we're probably good enough with these